### PR TITLE
Event manager

### PR DESCRIPTION
**Pull request by adam-westerman-bluefruit**
Implementation of a reusable event system.
Testing of basic use case

### CODE REVIEW

File: src/Event.c [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+#include <stdio.h>
+#include <string.h>
+
+#define MAX_LISTNERS 16 
+#define MAX_NAME_LENGTH 30
+
+struct Event 
+{
+    void (*fun_ptr_arr[MAX_LISTNERS])(void); // 8*16 = 128
+    uint8_t currentRegisteredListeners; // 8
```
Comment (martin-cooke-bluefruit) [inline src/Event.c]: Ideally, this kind of aid-memoir comment shouldn't appear in production code.
The pointer will be 32 bits, or 4 bytes, so the size in bytes of the struct is 4 * 16 + 1 = 65.

File: src/Event.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        mEventPointers[i] = &mEventStorage[i];
+    }
+    mNextUnusedEvent = 0;
+}
+
+EventPtr Event_createNew()
+{
+    if (mNextUnusedEvent >= MAX_EVENTS)
+    return 0; // no more free events
```
Comment (martin-cooke-bluefruit) [inline src/Event.c]: missing indent

File: src/Event.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            event->fun_ptr_arr[i+1] = NULL;
+        } 
+    }
+    event->currentRegisteredListeners--; 
+    return 0;
+}
+
+void Event_wipeAllListener(EventPtr event)
+{
+    if(event->currentRegisteredListeners == 0)return;
```
Comment (martin-cooke-bluefruit) [inline src/Event.c]: This check is identical to the one at the start of the for loop, so you could remove this one

File: src/Event.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    //
```
Comment (martin-cooke-bluefruit) [inline src/Event.c]: empty comment

File: src/Event.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+#include <stdint.h>
+
+typedef struct Event* EventPtr;
+
+void Events_system_init(void);
+void Event_invoke(EventPtr event);
+uint32_t Event_addListener(EventPtr event, void (*ptr)(void));
+uint32_t Event_removeListener(EventPtr event, void (*ptr)(void));
+void Event_wipeAllListener(EventPtr event);
+uint32_t Event_internal_getElementIndex(EventPtr event, void (*ptr)(void));
```
Comment (martin-cooke-bluefruit) [inline src/Event.h]: If the `internal` here implies that this function is only used inside of the Event module then you should not expose it here in the header and instead declare it as a static function at the top of the .c file

File: src/Event.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+#include <stdbool.h>
+#include <stdint.h>
+
+typedef struct Event* EventPtr;
+
+void Events_system_init(void);
+void Event_invoke(EventPtr event);
+uint32_t Event_addListener(EventPtr event, void (*ptr)(void));
+uint32_t Event_removeListener(EventPtr event, void (*ptr)(void));
+void Event_wipeAllListener(EventPtr event);
```
Comment (martin-cooke-bluefruit) [inline src/Event.h]: `wipeAllListener` is an unusual function to expose on the event system interface. You might have, say, system-A listening for an event from system-X and also system-B listening for the same system-X event. The value of the event system is decoupling, i.e. systems A, B, and X do not need to know anything about each other. Here you are allowing, for instance, system-A to wipe all listeners and potentially break system B's functionality.  There is likely a better way to design around this behaviour.

**Review by martin-cooke-bluefruit — COMMENTED** [review]
Nice work. you could generally remove some comments where the code is reasonably self-explanatory. It's easy for people to change the behaviour of the code but not update the comments and, of course, the compiler won't check your comments so they can easily go stale. Generally it's better to try to use descriptive function and variable names.

File: src/Methods.h [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#ifndef _Methods_H_
+#define _Methods_H_
+
+void Method_step_a();
+void Method_step_b();
+void Method_step_c();
+void Method_reset();
+int get_val_a();
+int get_val_b();
+int get_val_c();
```
Comment (martin-cooke-bluefruit) [inline src/Methods.h]: Presumably these helper functions are just for testing? If so, they should be in the `test` folder and not the `src` folder.

File: src/Event.h (hunk 6 above)
Comment (adam-westerman-bluefruit) [reply src/Event.h]: That is a good point. I'd designed this functionality for the purpose of wiping everything between tests, but i'll admit in the real world i've never designed an event pattern where the event cleans itself of listners. 

I think a more suitable method of cleaning is to simply remove the listeners which i've added to the test 1 for 1

File: src/Methods.h (hunk 7 above)
Comment (adam-westerman-bluefruit) [reply src/Methods.h]: Bingo. in the test folder they go

File: src/Event.h (hunk 6 above)
Comment (adam-westerman-bluefruit) [reply src/Event.h]: I've realised it's still useful for resetting an event when 'deleting', but definately shouldnt be public. I've changed it into a static function

File: src/Event.h (hunk 5 above)
Comment (adam-westerman-bluefruit) [reply src/Event.h]: I've made this static now. Do we have a common practice for naming static functions? I'm guessing '_internal' could be overexplaining something which clearly has the static keyword

File: src/Event.c [hunk 8]
```diff
@@ -7,18 +7,20 @@
...
-static struct Event mEventStorage[MAX_EVENTS]; // 136 * 32 = 4,352 ?
-static EventPtr mEventPointers[MAX_EVENTS]; // 8*32
+static struct Event mEventStorage[MAX_EVENTS];
+static EventPtr mEventPointers[MAX_EVENTS];
 static uint8_t mNextUnusedEvent;
 
-// allocate some space to store deleted
-static size_t pointerStorage;
+static bool Event_internal_getElementIndex(EventPtr event, void (*ptr)(void), int* index);
+static void Event_internal_wipeAllListeners(EventPtr event);
```
Comment (martin-cooke-bluefruit) [inline src/Event.c]: The convention we typically use is `Event_` implies that it's part of the public interface, and anything without that is an internal helper function. So just `static bool getElementIndex(EventPtr event, void (*ptr)(void), int* index);`

**Review by martin-cooke-bluefruit — APPROVED** [review]

**Pull request merged**